File: artifacts/tulipindicators/decompose/src/rsi.c
int ti_rsi_start(const double *options)
{
  return (int) options[0];
}


/*** DEPENDENCIES:
***/


int ti_rsi(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  const double per = 1.0 / ((double) period);
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_rsi_start(options))
  {
    return 0;
  }
  double smooth_up = 0;
  double smooth_down = 0;
  int i;
  for (i = 1; i <= period; i += 1)
  {
    const double upward = (input[i] > input[i - 1]) ? (input[i] - input[i - 1]) : (0);
    const double downward = (input[i] < input[i - 1]) ? (input[i - 1] - input[i]) : (0);
    smooth_up += upward;
    smooth_down += downward;
  }

  smooth_up /= period;
  smooth_down /= period;
  output[output_idx] = 100.0 * (smooth_up / (smooth_up + smooth_down));
  output_idx += 1;
  for (i = period + 1; i < size; i += 1)
  {
    const double upward = (input[i] > input[i - 1]) ? (input[i] - input[i - 1]) : (0);
    const double downward = (input[i] < input[i - 1]) ? (input[i - 1] - input[i]) : (0);
    smooth_up = ((upward - smooth_up) * per) + smooth_up;
    smooth_down = ((downward - smooth_down) * per) + smooth_down;
    output[output_idx] = 100.0 * (smooth_up / (smooth_up + smooth_down));
    output_idx += 1;
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_rsi_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_rsi_start(const double *options)
{
  return (int) options[0];
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/rsi.rs
use crate::*;
use std::assert;

pub fn ti_rsi_start(options: Option<&[f64]>) -> i32 {
    // Check if options is None (equivalent to NULL pointer check in C)
    // and if the slice has at least one element
    match options {
        Some(opt) if !opt.is_empty() => opt[0] as i32,
        _ => 0, // Default return value when options is None or empty
    }
}

pub fn ti_rsi(size: i32, inputs: Option<&[&[f64]]>, options: Option<&[f64]>, outputs: Option<&mut [&mut [f64]]>) -> i32 {
    // Unwrap options safely
    let options = match options {
        Some(opts) => opts,
        None => return 1,
    };
    
    // Get period from options
    let period = options[0] as i32;
    if period < 1 {
        return 1;
    }

    // Check if size is sufficient
    let start = ti_rsi_start(Some(options));
    if size <= start {
        return 0;
    }

    // Unwrap inputs and outputs safely
    let input = match inputs {
        Some(inp) => inp[0],
        None => return 1,
    };
    let output = match outputs {
        Some(out) => &mut out[0],
        None => return 1,
    };

    let per = 1.0 / (period as f64);
    let mut output_idx = 0;
    let mut smooth_up = 0.0;
    let mut smooth_down = 0.0;

    // Calculate initial smooth values
    for i in 1..=period {
        let upward = if input[i as usize] > input[(i - 1) as usize] {
            input[i as usize] - input[(i - 1) as usize]
        } else {
            0.0
        };
        let downward = if input[i as usize] < input[(i - 1) as usize] {
            input[(i - 1) as usize] - input[i as usize]
        } else {
            0.0
        };
        smooth_up += upward;
        smooth_down += downward;
    }

    smooth_up /= period as f64;
    smooth_down /= period as f64;
    output[output_idx] = 100.0 * (smooth_up / (smooth_up + smooth_down));
    output_idx += 1;

    // Calculate remaining values
    for i in (period + 1)..size {
        let upward = if input[i as usize] > input[(i - 1) as usize] {
            input[i as usize] - input[(i - 1) as usize]
        } else {
            0.0
        };
        let downward = if input[i as usize] < input[(i - 1) as usize] {
            input[(i - 1) as usize] - input[i as usize]
        } else {
            0.0
        };
        smooth_up = ((upward - smooth_up) * per) + smooth_up;
        smooth_down = ((downward - smooth_down) * per) + smooth_down;
        output[output_idx] = 100.0 * (smooth_up / (smooth_up + smooth_down));
        output_idx += 1;
    }

    // Verify output length
    assert!((output.as_ptr() as usize + output_idx * std::mem::size_of::<f64>()) - output.as_ptr() as usize
            == ((size - start) as usize * std::mem::size_of::<f64>()));

    0
}
